library/PracticeManager.js

```diff
@@ -0,0 +1,68 @@
+const moment = require('moment')
+const EventEmitter = require('events')
+
+/*
+ * The PracticeManager is intended to be a class that keeps
+ * track of the total time that a user has practiced in each guild. There are
+ * three kinds of times:
+ * - current time: increments when actively practicing
+ * - session time: sum of all time practiced between session resets (e.g., weekly)
+ * - overall time: sum of all time practiced for all time
+ *
+ * The PracticeManager does not handle persistence or role assignment.
+ * Instead it emits events that others can handle if they wish.
+ */
+class PracticeManager extends EventEmitter {
```

Comment:
This looks great. Just for my own clarification - the intent is that this class would call into the persistence layer, correct? In other words, somewhere in stopPractice() we'd have to reach into persistence to actually store the values? I assume we wouldn't be using the event for that purpose.

Reply:
Actually in the original PR I see

> Make a domain-only PracticeManager that is divorced from Discord events or persistence.

So ... is PracticeManager *not* intended to call into persistence? In that case, does some other driver reach into persistence and instantiate the PracticeManager with data already populated?

Reply:
I have been debating how this should work internally.

One option definitely be to inject a repository directly into `PracticeManager`, and then `stopPractice` could persist the updated times directly. Similarly, it would have to read times out of the repository instead of storing them in a Map.

Another option I was considering was to use a listener on the stopPractice event as an indicator to update the persistent store. And then, yes, we'd need some sort of factory that could pre-populate the manager with session and overall times at instantiation time.

Do you have a preference?

Reply:
I'm fine with the first option. I'd say decoupling the repository from PracMan (oh that's a fun abbreviation; I think I'm going to start using that label from now on) is probably design overkill.

Reply:
Yeah, probably right.